router/http.go

```diff
@@ -356,310 +345,27 @@ type httpService struct {
-	}
-	return conn, nil
-}
-
-type dialErr struct {
-	error
-}
-
-func (s *httpService) forwardAndProxyTCP(w http.ResponseWriter, req *http.Request, addrs []string, stickyAddr string, isSticky bool) {
-	var (
-		backend string
-		err     error
-		upconn  net.Conn
-	)
-	for _, backend = range addrs {
-		req.URL.Host = backend
-		upconn, err = dialer.Dial("tcp", req.URL.Host)
-		if err != nil {
-			// retry, maybe log a message about it
-			continue
-		}
-		defer upconn.Close()
-		break
-	}
-	if upconn == nil {
-		log.Println("no backends available")
-		failAndClose(w, 503)
-		return
-	}
-
-	err = req.Write(upconn)
-	if err != nil {
-		log.Println("error copying request to target:", err)
-		failAndClose(w, 503)
-		return
-	}
-
-	// Need to complete the handshake and set sticky cookie on response, otherwise
-	// websocket reconnections won't go to the right backend
-	upconnbr := bufio.NewReader(upconn)
-	res, err := http.ReadResponse(upconnbr, req)
-	if err != nil {
-		log.Println("http: proxy error:", err)
-		failAndClose(w, 503)
-		return
-	}
-	defer res.Body.Close()
-
-	// Copy the response headers and body over to the downstream ResponseWriter,
-	// the same as done in the non-TCP path.
-	copyHeader(w.Header(), res.Header)
-
-	if res.StatusCode != 101 {
-		// Upgrade was not successful, not going to reuse this connection.
-		w.Header().Set("Connection", "close")
-	}
-
-	if isSticky && stickyAddr != backend {
-		http.SetCookie(w, s.newStickyCookie(backend))
-	}
-
-	w.WriteHeader(res.StatusCode)
-	_, err = io.Copy(w, res.Body) // TODO(bgentry): consider using a flush interval
-	if err != nil {
-		log.Println("reverse proxy copy err:", err)
-		return
-	}
-
-	if res.StatusCode != 101 {
-		return
-	}
-	res.Body.Close() // close this now since we've copied everything
-
-	downconn, _, err := w.(http.Hijacker).Hijack()
-	if err != nil {
-		log.Println("hijack failed:", err)
-		failAndClose(w, 500)
-		return
-	}
-	defer downconn.Close()
-
-	errc := make(chan error, 2)
-	cp := func(dst io.Writer, src io.Reader) {
-		_, err := io.Copy(dst, src)
-		errc <- err
-	}
-	go cp(upconn, downconn)
-	go cp(downconn, upconnbr)
-	<-errc
-}
-
-func copyHeader(dst, src http.Header) {
-	for k, vv := range src {
-		for _, v := range vv {
-			dst.Add(k, v)
-		}
-	}
-}
-
-// Hop-by-hop headers. These are removed when sent to the backend.
-// http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html
-var hopHeaders = []string{
-	"Connection",
-	"Keep-Alive",
-	"Proxy-Authenticate",
-	"Proxy-Authorization",
-	"Te", // canonicalized version of "TE"
-	"Trailers",
-	"Transfer-Encoding",
-	"Upgrade",
-}
-
-type writeCloser interface {
-	CloseWrite() error
-}
-
-func shuffle(s []string) []string {
-	for i := len(s) - 1; i > 0; i-- {
-		j := random.Math.Intn(i + 1)
-		s[i], s[j] = s[j], s[i]
-	}
-	return s
+func isWebSocketUpgrade(req *http.Request) bool {
+	return strings.ToLower(req.Header.Get("Connection")) == "upgrade" &&
+		strings.ToLower(req.Header.Get("Upgrade")) == "websocket"
```

Comment:
Ahh, ok. I'll switch it to detect generic upgrade request here and in `ReverseProxy.ServeWebSocket`.